deploy/ethereum-GNOUSD/median/src/median.sol

```diff
@@ -0,0 +1,186 @@
+        return ecrecover(
+            keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", keccak256(abi.encodePacked(val_, age_, wat)))),
+            v, r, s
+        );
+    }
+
+    function poke(
+        uint256[] calldata val_, uint256[] calldata age_,
+        uint8[] calldata v, bytes32[] calldata r, bytes32[] calldata s) external
+    {
+        require(val_.length == bar, "Median/bar-too-low");
+
+        uint256 bloom = 0;
+        uint256 last = 0;
+        uint256 zzz = age;
+
+        for (uint i = 0; i < val_.length; i++) {
+            // Validate the values were signed by an authorized oracle
+            address signer = recover(val_[i], age_[i], v[i], r[i], s[i]);
+            // Check that signer is an oracle
+            require(orcl[signer] == 1, "Median/invalid-oracle");
+            // Price feed age greater than last medianizer age
+            require(age_[i] > zzz, "Median/stale-message");
+            // Check for ordered values
+            require(val_[i] >= last, "Median/messages-not-in-order");
+            last = val_[i];
+            // Bloom filter for signer uniqueness
+            uint8 sl = uint8(uint256(signer) >> 152);
+            require((bloom >> sl) % 2 == 0, "Median/oracle-already-signed");
+            bloom += uint256(2) ** sl;
+        }
+
+        val = uint128(val_[val_.length >> 1]);
+        age = uint32(block.timestamp);
+
+        emit LogMedianPrice(val, age);
+    }
+
+    function lift(address[] calldata a) external note auth {
+        for (uint i = 0; i < a.length; i++) {
+            require(a[i] != address(0), "Median/no-oracle-0");
+            uint8 s = uint8(uint256(a[i]) >> 152);
+            require(slot[s] == address(0), "Median/signer-already-exists");
+            orcl[a[i]] = 1;
+            slot[s] = a[i];
+        }
+    }
+
+    function drop(address[] calldata a) external note auth {
+       for (uint i = 0; i < a.length; i++) {
+            orcl[a[i]] = 0;
+            slot[uint8(uint256(a[i]) >> 152)] = address(0);
+       }
+    }
+
+    function setBar(uint256 bar_) external note auth {
+        require(bar_ > 0, "Median/quorum-is-zero");
+        require(bar_ % 2 != 0, "Median/quorum-not-odd-number");
+        bar = bar_;
+    }
+
+    function kiss(address a) external note auth {
+        require(a != address(0), "Median/no-contract-0");
+        bud[a] = 1;
+    }
+
+    function diss(address a) external note auth {
+        bud[a] = 0;
+    }
+
+    function kiss(address[] calldata a) external note auth {
+        for(uint i = 0; i < a.length; i++) {
+            require(a[i] != address(0), "Median/no-contract-0");
+            bud[a[i]] = 1;
+        }
+    }
+
+    function diss(address[] calldata a) external note auth {
+        for(uint i = 0; i < a.length; i++) {
+            bud[a[i]] = 0;
+        }
+    }
+}
+
+contract MedianGNOUSD is Median {
```

Comment:
this is the only part that really specifies the specific median. note that this is the pattern as defined in the canonical Maker docs